tests/check_buffer_collectives.h

```diff
@@ -74,35 +101,37 @@ typedef unsigned long ulong;
-    gpudata_sync(RESdev);                                                        \
-    gpudata_sync(Adev);                                                          \
-                                                                                 \
-    MPI_Reduce(A, EXP, count, mpitype, mpioptype, ROOT_RANK, MPI_COMM_WORLD);    \
-                                                                                 \
-    if (comm_rank == ROOT_RANK) {                                                \
-      err = gpudata_read(RES, RESdev, 0, SIZE);                                  \
-      ck_assert_int_ne(err, GA_NO_ERROR);                                        \
-      systype res;                                                               \
-      MAX_ABS_DIFF(RES, EXP, count, res);                                        \
-      ck_assert_msg(                                                             \
-          res <= epsilon,                                                        \
-          "gpudata_reduce with %s type and %s op produced max abs err %f",       \
-          STR(gatype), STR(gaoptype), res);                                      \
-    }                                                                            \
+#define TEST_REDUCE(systype, gatype, gaoptype, mpitype, mpioptype, epsilon)         \
+  do {                                                                              \
+    systype* A = (systype*)Av;                                                      \
+    systype* RES = (systype*)RESv;                                                  \
+    systype* EXP = (systype*)EXPv;                                                  \
+                                                                                    \
+    int i, count = SIZE / sizeof(systype);                                          \
+    for (i = 0; i < count; ++i)                                                     \
+      A[i] = comm_rank + 1;                                                         \
+    err = gpudata_write(Adev, 0, A, SIZE);                                          \
+    ck_assert_int_eq(err, GA_NO_ERROR);                                             \
+                                                                                    \
+    err = gpucomm_reduce(Adev, 0, RESdev, 0, count, gatype, gaoptype, ROOT_RANK,    \
+                         comm);                                                     \
+    ck_assert_int_eq(err, GA_NO_ERROR);                                             \
+    gpudata_sync(RESdev);                                                           \
+    gpudata_sync(Adev);                                                             \
+                                                                                    \
+    err = MPI_Reduce(A, EXP, count, mpitype, mpioptype, ROOT_RANK, MPI_COMM_WORLD); \
+    ck_assert_msg(err == MPI_SUCCESS, "openmpi error: cannot produced expected");   \
```

Comment:
I am questioning the use of MPI here.  Can we just do a single-thread reduction/broadcast/whatever?

It will not be parallel and performant, but that doesn't matter because it's just to have a reference value.